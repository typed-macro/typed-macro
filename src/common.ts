import { File, Program, Node } from '@babel/types'
import traverse, { NodePath } from '@babel/traverse'

export const nodeLoc = (node: Node) =>
  node.loc
    ? `line ${node.loc.start.line}, column ${node.loc.start.column}`
    : 'unknown'

export const validateFnName = (name: string) =>
  /^[_$a-zA-Z\xA0-\uFFFF][_$a-zA-Z0-9\xA0-\uFFFF]*$/.test(name)

export const findProgramPath = (ast: File) => {
  let path: NodePath<Program>
  traverse(ast, {
    Program(p) {
      path = p
      p.stop()
    },
  })
  return path!
}

/* eslint-disable @typescript-eslint/no-empty-function */
export const NO_OP = () => {}

export const findDuplicatedItem = <T>(a: T[], b: T[]) => {
  for (const item of a) {
    if (b.includes(item)) return item
  }
  return undefined
}
