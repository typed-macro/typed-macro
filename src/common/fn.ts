import { File, Program, Node } from '@babel/types'
import traverse, { NodePath } from '@babel/traverse'

export const nodeLoc = (node: Node) =>
  node.loc
    ? `line ${node.loc.start.line}, column ${node.loc.start.column}`
    : 'unknown'

export const validateFnName = (name: string) =>
  /^[_$a-zA-Z\xA0-\uFFFF][_$a-zA-Z0-9\xA0-\uFFFF]*$/.test(name)

export const findProgramPath = (ast: File) => {
  let path: NodePath<Program>
  traverse(ast, {
    Program(p) {
      path = p
      p.stop()
    },
  })
  return path!
}
