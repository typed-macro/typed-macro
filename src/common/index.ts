export * from './fn'
export * from './var'
