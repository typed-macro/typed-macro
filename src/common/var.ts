/* eslint-disable @typescript-eslint/no-empty-function */
export const NO_OP = () => {}
