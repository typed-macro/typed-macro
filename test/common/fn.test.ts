import { findProgramPath, nodeLoc, validateFnName } from '@/common/fn'
import { parse } from '@babel/parser'
import template from '@babel/template'
import { getAST } from '../testutils'

describe('common#nodeLoc', () => {
  it('should work', () => {
    const ast = getAST(`const a = 1`)
    expect(nodeLoc(ast.program.body[0])).toMatchSnapshot()
  })
})

describe('common#validateFnName', () => {
  it('should work', () => {
    const fnNames = [
      { name: '', ok: false },
      { name: '1a', ok: false },
      { name: 'a1', ok: true },
      { name: '你好世界', ok: true },
      { name: 'hello$world', ok: true },
      { name: '$yo', ok: true },
      { name: '_', ok: true },
    ]

    fnNames.forEach((item) => {
      expect(validateFnName(item.name)).toBe(item.ok)
    })
  })
})

describe('common#findProgramPath', () => {
  it('should work', () => {
    const ast = parse(`const a = 1`)
    const program = findProgramPath(ast)
    expect(program).not.toBeNull()
    expect(program).not.toBeUndefined()
  })

  it('should be synchronized when changed', () => {
    const ast = parse(`const a = 1`)
    const programA = findProgramPath(ast)
    const programB = findProgramPath(ast)
    programA.unshiftContainer('body', template.statement.ast('const b = 2'))
    expect(programA.get('body').length).toBe(2)
    expect(programB.get('body').length).toBe(2)
  })
})
