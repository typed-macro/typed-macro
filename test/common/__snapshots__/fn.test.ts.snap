// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common#nodeLoc should work 1`] = `"line 1, column 0"`;
